File: tests/nodejs/realworld_sqli.js
/**
 * Real-world SQL Injection test cases for Kunlun-M scanner
 * Tests: .query() with string concatenation, .raw() with template literal injection
 */

const express = require('express');
const knex = require('knex')({ client: 'mysql2', connection: {} });
const { Pool } = require('pg');
const pool = new Pool();
const router = express.Router();

router.use(express.json());
router.use(express.urlencoded({ extended: true }));

/**
 * VULN: SQL injection via knex.raw() with req.query string concatenation
 * Attacker can extract data via UNION or manipulate queries
 * Example: /users?search=' UNION SELECT password FROM admins --
 */
router.get('/users', async (req, res) => {
    const search = req.query.search;
    try {
        const results = await knex.raw(`SELECT * FROM users WHERE name LIKE '%${search}%'`);
        res.json(results);
    } catch (err) {
        res.status(500).send('Query failed');
    }
});

/**
 * VULN: SQL injection via pg Pool.query() with req.body string concatenation
 * Attacker can bypass authentication or extract sensitive data
 * Example POST: {"username": "admin' OR '1'='1", "password": "anything"}
 */
router.post('/login', async (req, res) => {
    const { username, password } = req.body;
    try {
        const result = await pool.query(
            `SELECT * FROM users WHERE username = '${username}' AND password = '${password}'`
        );
        if (result.rows.length > 0) {
            res.json({ success: true, user: result.rows[0] });
        } else {
            res.status(401).send('Invalid credentials');
        }
    } catch (err) {
        res.status(500).send('Login error');
    }
});

/**
 * VULN: SQL injection via knex.raw() with req.params concatenation
 * Attacker can inject SQL via the order parameter
 * Example: /products/desc; DROP TABLE products; --
 */
router.get('/products/:order', async (req, res) => {
    const order = req.params.order;
    try {
        const results = await knex.raw(`SELECT * FROM products ORDER BY price ${order}`);
        res.json(results);
    } catch (err) {
        res.status(500).send('Query failed');
    }
});

/**
 * SAFE: Parameterized query using knex binding syntax
 * User input is properly escaped and treated as a value, not SQL
 */
router.get('/safe-users', async (req, res) => {
    const search = req.query.search;
    try {
        // SAFE: Using parameterized query with bindings prevents SQL injection
        const results = await knex.raw('SELECT * FROM users WHERE name LIKE ?', [`%${search}%`]);
        res.json(results);
    } catch (err) {
        res.status(500).send('Query failed');
    }
});

/**
 * SAFE: Parameterized query using pg Pool with $1 placeholder
 * User input is safely passed as a parameter, not concatenated into SQL
 */
router.post('/safe-login', async (req, res) => {
    const { username, password } = req.body;
    try {
        // SAFE: Using $1/$2 parameterized placeholders prevents SQL injection
        const result = await pool.query(
            'SELECT * FROM users WHERE username = $1 AND password = $2',
            [username, password]
        );
        if (result.rows.length > 0) {
            res.json({ success: true, user: result.rows[0] });
        } else {
            res.status(401).send('Invalid credentials');
        }
    } catch (err) {
        res.status(500).send('Login error');
    }
});

module.exports = router;

File: tests/python/realworld_sqli.py
"""
Real-world SQL injection test cases for Kunlun-M scanner.
Simulates a Flask application with user management and search endpoints.
"""

import sqlite3
from flask import Blueprint, request, jsonify, g

users_bp = Blueprint('users', __name__)


def get_db():
    """Get database connection from Flask g object."""
    if 'db' not in g:
        g.db = sqlite3.connect('app.db')
        g.db.row_factory = sqlite3.Row
    return g.db


# ---------------------------------------------------------------------------
# VULN 1: SQL injection via string concatenation in SELECT
# ---------------------------------------------------------------------------
@users_bp.route('/api/users/search', methods=['GET'])
def search_users():
    """Search users by username."""
    keyword = request.args.get('q', '')
    db = get_db()
    # VULN: User input directly concatenated into SQL query
    # Attacker can inject: q="' OR 1=1 --" to dump all users
    query = f"SELECT id, username, email FROM users WHERE username LIKE '%{keyword}%'"
    cursor = db.execute(query)
    results = cursor.fetchall()
    return jsonify([dict(row) for row in results])


# ---------------------------------------------------------------------------
# VULN 2: SQL injection via f-string in UPDATE
# ---------------------------------------------------------------------------
@users_bp.route('/api/users/<int:user_id>/profile', methods=['POST'])
def update_profile(user_id):
    """Update user profile information."""
    bio = request.form.get('bio', '')
    db = get_db()
    # VULN: User input in f-string SQL without parameterization
    # Attacker can inject: bio="'; DROP TABLE users; --"
    query = f"UPDATE users SET bio = '{bio}' WHERE id = {user_id}"
    db.execute(query)
    db.commit()
    return jsonify({'status': 'updated'})


# ---------------------------------------------------------------------------
# VULN 3: SQL injection via string concatenation with ORDER BY
# ---------------------------------------------------------------------------
@users_bp.route('/api/users/list', methods=['GET'])
def list_users():
    """List users with sorting."""
    sort_col = request.args.get('sort', 'username')
    direction = request.args.get('dir', 'ASC')
    db = get_db()
    # VULN: User-controlled column name concatenated into ORDER BY clause
    # Attacker can inject: sort="username; DROP TABLE users --"
    query = f"SELECT id, username, email FROM users ORDER BY {sort_col} {direction}"
    cursor = db.execute(query)
    results = cursor.fetchall()
    return jsonify([dict(row) for row in results])


# ---------------------------------------------------------------------------
# SAFE 1: Parameterized query with ? placeholders
# ---------------------------------------------------------------------------
@users_bp.route('/api/users/<int:user_id>', methods=['GET'])
def get_user(user_id):
    """Get a single user by ID."""
    db = get_db()
    # SAFE: Parameterized query prevents SQL injection
    cursor = db.execute("SELECT id, username, email FROM users WHERE id = ?", (user_id,))
    user = cursor.fetchone()
    if user is None:
        return jsonify({'error': 'User not found'}), 404
    return jsonify(dict(user))


# ---------------------------------------------------------------------------
# SAFE 2: Parameterized query with named placeholders
# ---------------------------------------------------------------------------
@users_bp.route('/api/users/lookup', methods=['POST'])
def lookup_user():
    """Look up a user by email."""
    email = request.form.get('email', '')
    db = get_db()
    # SAFE: Using named parameter placeholders
    cursor = db.execute(
        "SELECT id, username, email FROM users WHERE email = :email",
        {'email': email}
    )
    user = cursor.fetchone()
    if user is None:
        return jsonify({'error': 'User not found'}), 404
    return jsonify(dict(user))
